filesystem/flysystem/FlysystemAssetStore.php

```diff
@@ -436,4 +735,91 @@ protected function getFileID($filename, $hash, $variant = null) {
 		return $fileID;
 	}
 
+	/**
+	 * Ensure each adapter re-generates its own server configuration files
+	 */
+	public static function flush() {
+		// Ensure that this instance is constructed on flush, thus forcing
+		// bootstrapping of necessary .htaccess / web.config files
+		$instance = singleton('AssetStore');
+		if ($instance instanceof FlysystemAssetStore) {
+			$publicAdapter = $instance->getPublicFilesystem()->getAdapter();
+			if($publicAdapter instanceof AssetAdapter) {
+				$publicAdapter->flush();
+			}
+			$protectedAdapter = $instance->getProtectedFilesystem()->getAdapter();
+			if($protectedAdapter instanceof AssetAdapter) {
+				$protectedAdapter->flush();
+			}
+		}
+	}
+
+	public function getResponseFor($asset) {
+		// Check if file exists
+		$filesystem = $this->getFilesystemFor($asset);
+		if(!$filesystem) {
+			return $this->createInvalidResponse();
+		}
+
+		// Block directory access
+		if($filesystem->get($asset) instanceof Directory) {
+			return $this->createDeniedResponse();
+		}
+
+		// Deny if file is protected and denied
+		if($filesystem === $this->getProtectedFilesystem() && !$this->isGranted($asset)) {
+			return $this->createDeniedResponse();
+		}
+
+		// Serve up file response
+		return $this->createResponseFor($filesystem, $asset);
+	}
+
+	/**
+	 * Generate an {@see SS_HTTPResponse} for the given file from the source filesystem
+	 * @param Filesystem $flysystem
+	 * @param string $fileID
+	 * @return SS_HTTPResponse
+	 */
+	protected function createResponseFor(Filesystem $flysystem, $fileID) {
+		// Build response body
+		// @todo: gzip / buffer response?
+		$body = $flysystem->read($fileID);
+		$mime = $flysystem->getMimetype($fileID);
+		$response = new SS_HTTPResponse($body, 200);
```

Comment:
@sminnee suggested looking at implementing http://www.php-fig.org/psr/psr-7/ at the same time.